### PR TITLE
lite-apiserver cache enhancement

### PR DESCRIPTION
**Pull request by Beckham007**
lite-apiserver cache enhancement, implement #51

### CODE REVIEW

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+	}
+
+	userAgent := getUserAgent(req)
+	key := cacheKey(userAgent, info)
+	klog.V(4).Infof("cache for %s", key)
+
+	return c.handleCache(key, info.Verb, statusCode, header, body)
+}
+
+func (c CacheManager) handleCache(key string, verb string, statusCode int, header http.Header, body io.ReadCloser) error {
```
[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/lite-apiserver/cache/cache_mgr.go:64]: Actually I'd like to have

```go
switch req.Method {
case "GET":
    c.cacheGet(xxx)
case "LIST":
    c.cacheList(xxx)
case "WATCH":
    c.cacheWatch(xxx)
}
```

Only these three methods are needed for cache.

File: pkg/lite-apiserver/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/superedge/superedge/pkg/lite-apiserver/constant"
+)
+
+// EdgeResponse hold all data of a response
+type EdgeCache struct {
+	Status int         `json:"code"`
```
[the remark this answers is not in the document]
Comment (dixudx) [reply pkg/lite-apiserver/cache/cache.go:12]: If the status code is not `200`, I think there is no need to do any cache.

If so, this field could be removed.

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+	info, ok := apirequest.RequestInfoFrom(req.Context())
+	if !ok {
+		return fmt.Errorf("parse requestInfo error")
+	}
+
+	userAgent := getUserAgent(req)
+	key := cacheKey(userAgent, info)
+	klog.V(4).Infof("cache for %s", key)
+
+	return c.handleCache(key, info.Verb, statusCode, header, body)
```
Comment (dixudx) [inline pkg/lite-apiserver/cache/cache_mgr.go:61]: Could ignore all non-200 responses.

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+		klog.V(6).Infof("new watch event, type=%s, obj=%v", eventType, obj)
+
+		switch eventType {
+		case watch.Bookmark:
+			klog.Infof("watch bookmark for key %s", key)
+		case watch.Error:
+			klog.Infof("watch error for key %s", key)
+		case watch.Added, watch.Modified, watch.Deleted:
+			listData, err := c.storage.LoadList(key)
```
Comment (dixudx) [inline pkg/lite-apiserver/cache/cache_mgr.go:109]: Maybe we could refactor codes in this `case`.

Add `continue` in above two cases to make below codes simpler and cleaner?

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+	return nil
+}
+
+func (c CacheManager) Query(req *http.Request) (*EdgeCache, error) {
+	info, ok := apirequest.RequestInfoFrom(req.Context())
+	if !ok {
+		return nil, fmt.Errorf("parse requestInfo error")
+	}
+
+	userAgent := getUserAgent(req)
```
Comment (dixudx) [inline pkg/lite-apiserver/cache/cache_mgr.go:230]: No need to check `userAgent` here

File: pkg/lite-apiserver/options/server_options.go [hunk 6]
```diff
@@ -29,13 +27,15 @@ type RunServerOptions struct {
...
 	CertFile          string
 	KeyFile           string
 	ApiserverCAFile   string
+	CacheType         string
 	FileCachePath     string
+	BadgerCachePath   string
+	BoltCacheFile     string
 }
 
 func NewRunServerOptions() *RunServerOptions {
```
Comment (dixudx) [inline pkg/lite-apiserver/options/server_options.go:41]: Could define default values in this function.

File: pkg/lite-apiserver/cache/cache.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package cache
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/superedge/superedge/pkg/lite-apiserver/constant"
+)
+
+// EdgeResponse hold all data of a response
```
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache.go]: done

File: pkg/lite-apiserver/cache/cache.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+func NewEdgeCache(status int, header http.Header, body []byte) *EdgeCache {
+	return &EdgeCache{
+		Status: status,
+		Header: header,
+		Body:   body,
+	}
+}
+
+func NewEmptyEdgeCache() *EdgeCache {
```
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache.go]: done, thx

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+func NewCacheManager(storage storage.Storage) *CacheManager {
+	return &CacheManager{
+		storage: storage,
+	}
+}
+
+func (c CacheManager) Cache(req *http.Request, statusCode int, header http.Header, body io.ReadCloser) error {
+	info, ok := apirequest.RequestInfoFrom(req.Context())
+	if !ok {
+		return fmt.Errorf("parse requestInfo error")
```
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache_mgr.go]: done, thx

File: pkg/lite-apiserver/cache/cache_mgr.go (hunk 3 above)
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache_mgr.go:62]: we only cache Http.OK now, filter this in proxy.modifyResponse()

File: pkg/lite-apiserver/cache/cache_mgr.go (hunk 4 above)
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache_mgr.go:113]: added continue

File: pkg/lite-apiserver/cache/cache_mgr.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache_mgr.go:65]: `handleCache ` is used for test. I will add more test if time permits.

File: pkg/lite-apiserver/cache/cache_mgr.go [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+func cacheKey(userAgent string, info *apirequest.RequestInfo) string {
+	keys := []string{userAgent, info.Namespace, info.Resource, info.Name, info.Subresource}
+	key := strings.Join(keys, "_")
+	return key
+}
+
+func getUserAgent(r *http.Request) string {
+	userAgent := constant.DefaultUserAgent
```
[the remark this answers is not in the document]
Comment (Beckham007) [reply pkg/lite-apiserver/cache/cache_mgr.go:286]: done, thx

File: pkg/lite-apiserver/options/server_options.go (hunk 6 above)
Comment (Beckham007) [reply pkg/lite-apiserver/options/server_options.go:41]: default values define in `AddFlags`

**Pull request merged**